Outbox Collection

pfefferle opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

Working on an Outbox Collection, to better support Blogs with lots of Followers, Retries and Activities other than Post.

/cc @mediaformat @Menrath @ruru4143 @akirk @mattwiebe 

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Add Outbox
* Scheduler, that handles unpublished Outbox items in chunks
* Implement Retry
* Actions to push data into the Outbox

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Go to '..'
*

Fixes #961 #1203

pfefferle committed:
init

pfefferle committed:
outbox should not be public!

pfefferle committed:
add basic `to_activity` function

pfefferle committed:
Merge branch 'master' into add/outbox-collection

pfefferle committed:
do not allow to instance post transformer

pfefferle committed:
Merge branch 'master' into add/outbox-collection

pfefferle committed:
fix tests

pfefferle committed:
Merge branch 'master' into add/outbox-collection

pfefferle committed:
Merge branch 'master' into add/outbox-collection

pfefferle committed:
Merge branch 'master' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
fix phpcs

pfefferle committed:
revert change

pfefferle committed:
remove unused `use` declarations

pfefferle committed:
the handler should not handle outgoing stuff

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Update class-activitypub.php

pfefferle committed:
fix PHPCS

pfefferle committed:
more PHPCS fixes

pfefferle committed:
fix namespace issue

pfefferle committed:
remove unneeded function

pfefferle committed:
fix sticky post endpoint

/cc @mattwiebe

pfefferle committed:
no need to check for User-ID

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
support JSON and Arrays beside WP_Comments and WP_Posts

pfefferle committed:
fix auto-complete issue

pfefferle committed:
simplify code.

pfefferle committed:
convert JSON to Activity_Object to not have everything duplicated

pfefferle committed:
add support for maps and mentions

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

obenland commented on includes/transformer/class-activity-object.php:
```diff
@@ -0,0 +1,155 @@
+	 * @return array The list of Mentions.
+	 */
+	protected function get_cc() {
+		$cc       = array();
+		$mentions = $this->get_mentions();
+
+		if ( $mentions ) {
+			foreach ( $mentions as $url ) {
+				$cc[] = $url;
+			}
+		}
+
+		return $cc;
```
```suggestion
		return $this->get_mentions() ?: array();
```
Could this be simplified to ^?
If `get_mentions()` always returned an array, the fallback would not be needed.

pfefferle replied:
We can remove the `if` then and directly iterate but we can't simply return `get_mention()` because it has some more data, not only the list of URLs.

pfefferle replied:
And we could use an `array_filter` instead!

obenland replied:
Hm, maybe `return array_values( $this->get_mentions() );`? 

Currently the function doesn't remove anything from `$mentions`, it just puts everything in `$cc`.

pfefferle replied:
🤔

pfefferle replied:
You are right, array_values should do the trick! It thought it has some more informations, but it's simply the key is the WebFinger handle and the URL is the value!

pfefferle committed:
it takes a string (JSON) or an array

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
simplified based on the feedback of @obenland

pfefferle committed:
Merge branch 'add/outbox-collection' of https://github.com/Automattic/wordpress-activitypub into add/outbox-collection

pfefferle committed:
rearrange code a bit

pfefferle committed:
add missing uses

pfefferle committed:
fix typo

pfefferle committed:
add missing uses

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Added Changelog-Entry

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
apply changes from @obenland

pfefferle committed:
delay scheduling a bit

pfefferle committed:
change scheduling and add possibility to add items to outbox

pfefferle committed:
remove deprecated functions

pfefferle committed:
add filter

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
fix indents

pfefferle committed:
removed deprecated functions

pfefferle committed:
add visibility and normalize autor id

pfefferle committed:
add taxonomies to store actor and actitiy-type informations

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

obenland committed:
Merge branch 'trunk' into add/outbox-collection

obenland committed:
Fix phpcs

obenland committed:
Fix shortcode tests

obenland committed:
Update json tests.

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
move map functions to base transformer

pfefferle committed:
use item instead of wp_object

pfefferle committed:
protect functions

pfefferle committed:
fix tests and remove legacy tests

pfefferle committed:
fix phpcs

pfefferle committed:
add unittests

pfefferle committed:
fix attachment issues and add tests

pfefferle committed:
add tests for JSON transformer and fix detected problems

pfefferle committed:
fix phpcs issues

pfefferle committed:
add activity-object transformer tests

pfefferle committed:
simplify code

pfefferle committed:
define as global functions

pfefferle committed:
re-added setter

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
add content visibility

pfefferle committed:
fix PHPCS

obenland committed:
Outbox: Use post meta instead of taxonomies (#1173)

* Outbox: Use post meta instead of taxonomies

* Add tests for meta values when adding Outbox items

* fix indents

---------

Co-authored-by: Matthias Pfefferle <pfefferle@users.noreply.github.com>

mattwiebe commented on includes/functions.php:
```diff
@@ -1593,3 +1595,48 @@ function is_self_ping( $id ) {
+
+	set_wp_object_state( $data, 'federate' );
+
+	$id = Outbox::add( $activity, $type, $user_id, $content_visibility );
```
It seems to me that all of this functionality in `add_to_outbox` should just live inside of `Outbox::add`? Which would then either leave us with a thin wrapper in this function, or we can then get rid of the wrapper entirely?

mattwiebe commented on includes/scheduler/class-actor.php:
```diff
@@ -0,0 +1,102 @@
+	 * @return mixed
+	 */
+	public static function blog_user_update( $value = null ) {
+		self::schedule_profile_update( 0 );
```
```suggestion
		self::schedule_profile_update( Actors::BLOG_USER_ID );
```

As the person who wrote the original code and several times thought, "I should patch that"…

mattwiebe reviewed (commented):
Haven't tried running this yet, but read through the changes and left a couple of comments/questions.

Really like the scheduler separation. 👍 

What do we still need to do to ship this?

pfefferle committed:
Update includes/scheduler/class-actor.php

Co-authored-by: Matt Wiebe <wiebe@automattic.com>

pfefferle committed:
register content-visibility meta

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

obenland committed:
Outbox: Update API Endpoint (#1170)

* Update file name

* Extend WP_REST_Controller

* First pass at updated outbox endpoint

* Remove expectation of having cc in the response

* Update object type to be a bit more descriptive

* First pass at create endpoint

* Fix tests

* Get Activity Type from meta after #1173

* Fix rest_url_path

Props @pfefferle

* Return accurate number of outbox items

Props @pfefferle

* Add more tests and remove unnecessary out-of-bounds error

* Remove POST endpoint for now

We currently don't have a use-case for it.

* Limit Outbox by activity type and visibility

* Account for posts without visibility meta

Props @pfefferle

* phpcs

* Allow logged in users to see all activity types

* Add request context to remaining hooks

* Default query to limit activities

Adds an exception for requests with privileges.

* Slightly improve query

pfefferle committed:
fix unittests

pfefferle committed:
Merge branch 'add/outbox-collection' of https://github.com/Automattic/wordpress-activitypub into add/outbox-collection

pfefferle commented on includes/functions.php:
```diff
@@ -1547,3 +1549,48 @@ function is_self_ping( $id ) {
+			$args
+		);
+	}
+
```
```suggestion

	if ( 'Create' === $type && ACTIVITYPUB_CONTENT_VISIBILITY_PUBLIC && $content_visibility && ACTIVITYPUB_ACTOR_AND_BLOG_MODE === \get_option( 'activitypub_actor_mode', ACTIVITYPUB_ACTOR_MODE ) ) {
		// @todo transform $data to announce
		add_to_outbox( $data, 'Announce', Actors::BLOG_USER_ID );
	}
```

mattwiebe replied:
Would edits also need to be supported, or do those merely reference the original post? If so might we need to send those direct post updates to followers of the Blog Actor?

pfefferle replied:
you only share the id of a post via an `Announce`, so only `Creates` are important here... the updates will be done automatically afterwards... no need to re-announce them...

mattwiebe replied:
> you only share the id of a post via an Announce, so only Creates are important

Perfect. But the further question:

Let's say I follow a WP `example.com` site from `mattwiebe@mastodon.social` with `ACTIVITYPUB_ACTOR_AND_BLOG_MODE === true`. I follow `example.com@example.com`, but I don't follow `user@example.com`. I receive the `Announce` of a post by `user@example.com`. But the user then edits the post. Will `mastodon.social` get notified of the edit with nobody on the instance following `user@example.com`?

pfefferle replied:
🤔 good point...

mattwiebe replied:
So this means that we'll need to also POST to the inboxes of the Blog Actor for`Update`s. Are there any other actions you can think of that should be included in that behaviour?

I'll add a commit to #1186 to do that.

pfefferle replied:
I would start with these two

pfefferle replied:
(and I am still not sure that update needs to be federated)

pfefferle replied:
the current code supports `Create`, `Update` and `Delete` https://github.com/Automattic/wordpress-activitypub/blob/1f7abb76ef4fa171ec7d120eaa03eac3913d3c78/includes/class-activity-dispatcher.php#L104

mattwiebe replied:
Perfect. I did `Update` but I added `Delete` as well - https://github.com/Automattic/wordpress-activitypub/pull/1186/commits/2caab9bdd6f7488f944fce5b6a132b0413f298d8

pfefferle committed:
pending seems to be the better status

pfefferle committed:
Revert "pending seems to be the better status"

This reverts commit 8a7b5ccbf3f55bc5ce85b04e7366e214a09d32e0.

Menrath commented on includes/class-activitypub.php:
```diff
@@ -518,6 +519,102 @@ private static function register_post_types() {
+			Outbox::POST_TYPE,
+			'_activitypub_activity_actor',
+			array(
+				'type'              => 'integer',
```
Shouldn't the type be string?

mattwiebe committed:
Merge branch 'trunk' into add/outbox-collection

Menrath commented on includes/collection/class-outbox.php:
```diff
@@ -0,0 +1,75 @@
+	 */
+	public static function add( $activity_object, $activity_type, $user_id, $content_visibility = ACTIVITYPUB_CONTENT_VISIBILITY_PUBLIC ) { // phpcs:ignore
+		switch ( $user_id ) {
+			case -1:
```
Shouldn't the constants `Activitypub\Collection\Actors::BLOG_USER_ID` and `Activitypub\Collection\Actors::APPLICATION_USER_ID` be used?

pfefferle replied:
Addressed here: https://github.com/Automattic/wordpress-activitypub/pull/1200/commits/4d9f1a40f6401a16db2ccc19e5e7b4eeeb036860

Menrath commented on includes/collection/class-outbox.php:
```diff
@@ -0,0 +1,75 @@
+namespace Activitypub\Collection;
+
+/**
+ * ActivityPub Outbox Collection
```
```suggestion
 * ActivityPub Outbox Collection
 *
 * @link https://www.w3.org/TR/activitypub/#outbox
```

Menrath commented on includes/collection/class-outbox.php:
```diff
@@ -0,0 +1,75 @@
+	/**
+	 * Add an Item to the outbox.
+	 *
+	 * @param \Activitypub\Activity\Base_Object $activity_object    The Activity-Object  to add as JSON.
```
```suggestion
	 * @param \Activitypub\Activity\Base_Object $activity_object    The object of the activity that will be added to the outbox.
```

Menrath commented on includes/collection/class-outbox.php:
```diff
@@ -0,0 +1,75 @@
+	 *
+	 * @param \Activitypub\Activity\Base_Object $activity_object    The Activity-Object  to add as JSON.
+	 * @param string                            $activity_type      The activity type.
+	 * @param int                               $user_id            The user ID.
```
The real or imaginary user ID of the actor that published the activity that will be added to the outbox.
```suggestion
	 * @param int                               $user_id            The real or imaginary user ID of the actor that published the activity that will be added to the outbox.
```

Menrath commented on includes/collection/class-outbox.php:
```diff
@@ -0,0 +1,75 @@
+		$outbox_item = array(
+			'post_type'    => self::POST_TYPE,
+			'post_title'   => $activity_object->get_id(),
+			'post_content' => $activity_object->to_json(),
```
Maybe always using `post_content_filtered` to store the JSON could make the beneath removal of filter needless.

Menrath commented on includes/collection/class-outbox.php:
```diff
@@ -0,0 +1,75 @@
+			'post_type'    => self::POST_TYPE,
+			'post_title'   => $activity_object->get_id(),
+			'post_content' => $activity_object->to_json(),
+			// ensure that user ID is not below 0.
```
It would be helpful to know why this has to be ensured.

pfefferle replied:
because WP_Post does not support negative ints

Menrath commented on includes/functions.php:
```diff
@@ -1547,3 +1549,48 @@ function is_self_ping( $id ) {
+		return false;
+	}
+
+	set_wp_object_state( $data, 'federate' );
```
The namings within the function set_wp_object_state should probably be updated as well.

pfefferle replied:
this is still only for WP_Objects and might be deprecated over time

obenland committed:
Outbox: Use callbacks that can be unhooked. (#1188)

* Outbox: Use callbacks that can be unhooked.

* Add unit tests

* Use data provider in no-activity tests

* Make sure user has activitypub cap

* Use should_be_federated() instead

See #1186.

* Update with learnings from #1196

obenland committed:
Outbox: Make sure Last page is linked correctly (#1195)

Props @pfefferle

Menrath commented on includes/scheduler/class-post.php:
```diff
@@ -0,0 +1,113 @@
+		}
+
+		// Check if post-type supports ActivityPub.
+		$post_types = \get_post_types_by_support( 'activitypub' );
```
Wouldn't it be better if this check is in the `is_post_disabled` function? Because that one allows third parties to filter. If I want to disable a post type in general, but have one I want to still federate, then I also need to take care of my own scheduler.

Menrath reviewed (commented):
I have looked through that code, and am really looking forward to see this in production. I made some comments, but didn't find anything severe.

I will test my Event Bridge for ActivityPub plugin against this PR in the next days and maybe have some more feedback then. 

Keep on the amazing work!

pfefferle committed:
Update includes/collection/class-outbox.php

Co-authored-by: André Menrath <99024746+Menrath@users.noreply.github.com>

pfefferle committed:
Update includes/collection/class-outbox.php

Co-authored-by: André Menrath <99024746+Menrath@users.noreply.github.com>

pfefferle committed:
Update includes/collection/class-outbox.php

Co-authored-by: André Menrath <99024746+Menrath@users.noreply.github.com>

pfefferle committed:
fix type

props @menrath

pfefferle committed:
Outbox: Fix Query (#1200)

* Outbox: Fix Query

If `author` is null, it will return every activity for the blog and application Actor, but we only want to have the Actors activities.

* Update includes/rest/class-outbox-controller.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

* fix user check!

* fix phpcs

* use constants

* Add tests for new functionality

* Update tests/includes/rest/class-test-outbox-controller.php

---------

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Fix typo

obenland committed:
Outbox: Tests and fixes for Post Scheduler (#1196)

* Outbox: Tests and fixes for Post Scheduler

* Move extrafields post to Actor class

* Add test coverage for remaining Actor methods

* Move cap-manipulating tests to the end

* Account for inheritance in Transformer Factory

Reverts changes to `add_to_outbox()`

* Remove outbox items after every test.

* Use fully qualified class name

* Split extra fields tests to not hit time limit

* Adjust bog modes

* debug

* debug

* debug

* debug

* Use Profile_update action.

`wp_update_user` wasn't introduced until WP 6.3.

* revert changes to add_to_outbox

pfefferle committed:
Update Dispatcher to use Outbox (#1186)

* rename dispatcher

* Simple dispatcher based on the Outbox-Collection

This is a simple rewrite of the current dispatcher system, to use the Outbox instead of the Scheduler.

This is a first draft and will be improved over time, to better handle:

* Re-tries
* Errors
* Logging
* Batch processing

* update changelog

* mark post as `publish` after federation id done

* show only published activities

* fix missing rename

* use pending instead of draft

* do not check for post_status

* fix tests

props @obenland

* Send `Update`s to Blog Actor in dual mode

* Update includes/class-dispatcher.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

* Update includes/class-dispatcher.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

* Update includes/rest/class-outbox-controller.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

* Check if Activity should be sent to followers

* the unique check will be done `send_activity_to_followers`

* fix tests

* fix PHPCS

* move scheduler behind action

* Add `private` visibility

* Add Announce activity

* Announce the full object!

* fix indent

* Update includes/transformer/class-base.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

* add doc-block

* only boost content not profile updates

* Also handle `Delete` when bundling Blog Actor inboxes

* Update docs

* Avoid activitypub_actor_mode bleeding into other tests

* Fix comments tests

* Account for inheritance in Activity objects

* Move hook to the right place

* fix typo!

* trigger scheduler

* Fix tests

---------

Co-authored-by: Matt Wiebe <wiebe@automattic.com>
Co-authored-by: Konstantin Obenland <obenland@gmx.de>

obenland committed:
Outbox: Align return type and update docs (#1211)

obenland committed:
Outbox: Fix broken HTML tags after decoding Outbox items (#1212)

pfefferle committed:
Outbox: Fix race condition with Post-Metas (#1214)

* Outbox: Fix race condition with Post-Metas

* Update includes/scheduler/class-post.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

---------

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Outbox: Generalize functionality of the transformers (#1178)

* Generalize functionality of the transformers

To better support

* extraction of mentions for arbitrary data
* generating the audience based on the content visibility
* ...

* use `cc` or `to` if set and fall back to public

* revert renaming

* oops!

* keep mentions class

* revert test class

* No need to transform it again!

* unify content visibility

* fix check

* no need to pass the visibility

* overwrite post visibility

* fix tests

* fix most of the issues

* fix test

* ensure that actor is set before sending the Activity!

* fix last tests

* Fix PHPCS

* add ID

* add a more descriptive title

* make function global

* fix issue if no URI is provided

* fix race condition

* fix spaces

* revert race-condition fix

See: https://github.com/Automattic/wordpress-activitypub/pull/1214

* revert title

* fix unittests

* fix content visibility for comments

* Update includes/transformer/class-base.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

---------

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

obenland committed:
Update/testcases (#1215)

* Outbox: Add testcase with common

* Remove unused post_id

* Fix test to actually test something

obenland committed:
Outbox: Add upgrade script to populate outbox items (#1175)

* Outbox: Add upgrade script to populate outbox items

* Don't mess with federation

* First pass at tests

* Use post fixtures

* Split out add_to_outbox method

* Split upgrade routines for posts and comments

* add unit tests

* Account for any post status in tests

* Account for publish meaning federated

* Skip posts from users who don't have the caps.

* Restrict outbox migration by `activitypub_status===federated` meta

* lint

* `any` post_type

props @obenland

* do not migrate outbox items. fix tests

* Account for an arbitrary number of args to be passed

* back to unreleased, add comment

* oops

* Update function docs and signature

* Account for disabled users

---------

Co-authored-by: Matt Wiebe <wiebe@automattic.com>

obenland committed:
Outbox: Fall back to blog author for non-authorized authors (#1208)

* Outbox: Don't create items for post from not activitypub-authorized authors

* Use utility function

Props @pfefferle

* Move user_id check into outbox helper

* Update unit tests

* Revert changes in post class

* Skip when blog mode is disabled

pfefferle committed:
Accept Follower: Migrate to outbox (#1205)

* Follower: Migrate to outbox

* add missing namespace

* remove unused dms/phpunit-arraysubset-asserts dependency

* add support for Activities

* add tests

* fix phpcs issues

* re-add dev-dependency

* always add actor to get sure it is the right one!

* Update includes/functions.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

* Update includes/functions.php

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

* simplify tests

* rename function

---------

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

obenland committed:
Send Activity and not its object (#1217)

* Create failing test

* Set activity object as an object.

* minimize changes

pfefferle committed:
Outbox: fix to_id function for user transformer (#1218)

* fix to_id function for user transformer

* fix phpcs

* Add test

---------

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

obenland committed:
Update class-test-outbox.php (#1220)

akirk commented on includes/functions.php:
```diff
@@ -1547,3 +1551,121 @@ function is_self_ping( $id ) {
+ * @param mixed   $data               The object to add to the outbox.
+ * @param string  $activity_type      The type of the Activity.
+ * @param integer $user_id            The User-ID.
+ * @param string  $content_visibility The visibility of the content.
```
Could you add the valid strings in the documentation here? Or where you can find them.

pfefferle replied:
I added the information

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Add hint where to find the visibility constants

pfefferle committed:
rename function to match new structure

props @obenland

obenland committed:
Outbox: Fix attachment transitions (#1219)

* Outbox: Correctly handle attachments

* Update function docs to reflect reality

* Aside: Add some context as to why this callback is necessary

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Outbox: Fix deletes (#1224)

* Outbox: Fix deletes

* Always send deletes

pfefferle committed:
Outbox: Basic rescheduling (#1223)

* Outbox: Basic rescheduling

* add tests

* Add tests

* reuse ActivityPub_Outbox_TestCase class

* fix phpcs issues

* update tests

* fix tests

* fix tests

* debug

* debug

* types?

* add missing hook

props @obenland

* simpified code

props @obenland

* remove unused namespace definitions

---------

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

obenland committed:
Outbox: Publish item after processing (#1225)

pfefferle committed:
Outbox: Fix `to`, `cc` and `id` (#1226)

props @obenland

obenland committed:
Outbox: Register new event schedule on upgrade (#1228)

obenland committed:
Outbox: Make post type non-hierarchical (#1227)

* Outbox: Make post type non-hierarchical

* Default to title and content

It helps when setting the post type to public to debug something

pfefferle committed:
Outbox: Fix profile updates (#1233)

obenland committed:
Outbox: Account for transformer errors (#1231)

* Outbox: Account for transformer errors

* Remove debug

obenland committed:
Outbox: Account for invalid json (#1230)

* Outbox: Account for invalid json

* http it is!

* Remove unnecessary init

pfefferle committed:
Merge branch 'trunk' into add/outbox-collection

pfefferle committed:
Outbox: Fix profile updates (#1236)

* Outbox: Fix profile updates

This will transform the stored JSON into the correct object!

* phpcs fix

obenland committed:
Outbox: Federate blog user updates (#1237)

* Outbox: Federate blog user updates

* Add right hooks and complete unit tests

obenland committed:
Outbox: Show Outbox processing in Stream (#1229)

* Outbox: Show Outbox processing in Stream

* Make sure we keep them all and not just the first one

* Switch to new action

* Keep track of inbox URL

obenland committed:
Merge branch 'trunk' into add/outbox-collection
